// ccusage/dist/calculate-cost-BDqO4yWA.js
import { getTotalTokens } from "./_token-utils-WjkbrjKv.js";
/**
* Calculates total token usage and cost across multiple usage data entries
* @param data - Array of daily, monthly, or session usage data
* @returns Aggregated token totals and cost
*/
function calculateTotals(data) {
	return data.reduce((acc, item) => ({
		inputTokens: acc.inputTokens + item.inputTokens,
		outputTokens: acc.outputTokens + item.outputTokens,
		cacheCreationTokens: acc.cacheCreationTokens + item.cacheCreationTokens,
		cacheReadTokens: acc.cacheReadTokens + item.cacheReadTokens,
		totalCost: acc.totalCost + item.totalCost
	}), {
		inputTokens: 0,
		outputTokens: 0,
		cacheCreationTokens: 0,
		cacheReadTokens: 0,
		totalCost: 0
	});
}
/**
* Creates a complete totals object by adding total token count to existing totals
* @param totals - Token totals with cost information
* @returns Complete totals object including total token sum
*/
function createTotalsObject(totals) {
	return {
		...totals,
		totalTokens: getTotalTokens(totals)
	};
}
export { calculateTotals, createTotalsObject };
